### PR TITLE
Strictly use C++14 only

### LINKED ISSUE #1296: C++17 contamination
#1291 introduced some C++17 features.
Since the requirement is set at C++14, we need to workaround these changes.

### PR DESCRIPTION
**Pull request by ye-luo**
To fix #1296.
To test AFQMC on Rhea, GNU 6 was used on the CPU build. Although GNU 6 is not fully C++17 compliant, some feature is already available and the use of C++17 feature went through with the current CI.
In order to avoid C++17 contamination, this PR put back GNU 5 on the CI CPU build.
Some fix is needed to workaround the use of C++17.

I made some obvious fix but further fix needs AFQMC experts. @fdmalone @mmorale3 Could you check the errors with GNU 5.

### CODE REVIEW

File: tests/test_automation/jenkins_rhea_cpu.sh [hunk 1]
```diff
@@ -16,7 +16,7 @@ cd $BUILD_DIR
 source /sw/rhea/environment-modules/3.2.10/rhel6.7_gnu4.4.7/init/bash
 
 module unload PE-intel
-module load PE-gnu/6.2.0-2.0.1
+module load PE-gnu/5.3.0-1.10.2
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply tests/test_automation/jenkins_rhea_cpu.sh]: cmake put -std=c++14 for us.
A C++17 code does conform C++14 standard and the compilation went through. I need a way to block C++17 features.

[the remark this answers is not in the document]
Comment (grahamlopez) [reply tests/test_automation/jenkins_rhea_cpu.sh]: Okay, I double-checked, and setting `-std=c++14` only produces warnings when C++17 features are used. Once we clean up all of the compiler warnings, we could set `-Werror` to turn these into compile errors, but I don't yet see another way to have g++ reject C++17 code.

[the remark this answers is not in the document]
Comment (ye-luo) [reply tests/test_automation/jenkins_rhea_cpu.sh]: Put back 5.3 probably is a practical way to stop the use of most C++17 and beyond. Currently the GPU CI is with 5.3, minimize the uncertainty in CI setting is my prefererence.

**Comment by prckent** [comment]
Good point to trap C++17 at this stage. Consistent with our testing with Boost 1.61.0, the minimum required Boost version, it would be useful to catch C++17 usage since we state C++14 is the minimum. If this can't be done easily, we could begin stating which compiler versions are acceptable and tested against.

**Comment by mmorale3** [comment]
This was an oversight on my end, I’ll fix it today. But I agree that blocking c++17 is a good practice. 

Miguel

**Comment by mmorale3** [comment]
I can't open the link to rhea, do I need special permission or access?

**Comment by grahamlopez** [comment]
> I can't open the link to rhea, do I need special permission or access?

No, the way it is set up you have to be connecting from within the network. I'm working on getting email notifications to send out the build log to PR contributors, but it is taking some time

**Comment by mmorale3** [comment]
Ye, when I run ctest manually in rhea all the tests are passing. 
Now I would need to see the test output to figure out what the problem is.

**Comment by prckent** [comment]
OK to test

**Comment by grahamlopez** [comment]
test this please

**Comment by ye-luo** [comment]
@mmorale3 Thank you for all the fix. Everything is good. This PR is ready to merge.

**Pull request merged**